### PR TITLE
Refactoring of configuration

### PR DESCRIPTION
**Pull request by pykong**
This **PR** cleans and hardens the code base via the following refactorizations:
- Untangling the creation of the configuration object from `copy_local`.
- Introducing basic data validation and default handling via `pydantic`.
- Various minor improvements like making `flags` a class for explicit member access via dot syntax.

#### TODO
- [x] Make tests green.
    - [x] Add `pydantic` to dependencies.
    - [x] Fix type annotations.
- [x] Consider moving `user_data` into sub module to together with `conf`
- [x] General cleanup and removal of dead code.
- [x] Add new tests.
   - [x] Test merging precedence.

Preceded by: #54

Closes: #49

### CODE REVIEW

**Comment by pykong** [comment]
@jpscaletti 

I am excited to announce that with e2b4742 this **PR** has now become merge ready. 

I have put a lot of work and love into making `copier`'s code shine. The result is a code that I believe is a configuration handling that is much DRYer, easier to maintain and better structured. I really hope you like what I have created and are going to merge this **PR**.

As the changes to `copier`'s code are quite substantial, I would appreciate if you could perform a basic code review.

**Comment by pykong** [comment]
I am going to write some additional tests, too.

**Comment by pykong** [comment]
With 22d165e I have noe added some primitive tests for the three `pydantic` models to ensure they validate, transform and load input correctly.

**Comment by pykong** [comment]
I have further polished this **PR** beyond what I deemed a first merge candidate.
With 3b5245c it has now reached a stage at which I am not going to add any further commits, except if demanded by a code review.

**Comment by pykong** [comment]
@jpscaletti Could you maybe have a look into this PR so we can move `copier` forward?

**Comment by pykong** [comment]
@jpscaletti 
You started to resolve the merge conflicts and perform additional changes. Do you require me to do anything to move this **PR** forward?

**Comment by pykong** [comment]
I have merged `master/origin` into `better_conf` (the branch behind this PR).
Surprisingly, this operation has not resolved the merge conflicts.

Additionally, **Travis** did not run on import copier

I am going to investigate how to persuade GitHub to do our bidding here.

**Comment by pykong** [comment]
@jpscaletti Would you mind granting me write access to this repo, as this would allow me to more easily resolve the conflicts? Right now I can not even inspect on GitHub what the actual conflicts are.

image

**Comment by jpscaletti** [comment]
Ok. But you might have more luck doing a `git rebase`. The conflicts seem to be caused by my changes from `from ..copier import x` to `from copier import x` in the tests files (done when removing the root __init__ file)

**Comment by pykong** [comment]
@jpscaletti That's strange I have already fixed the exact same issues you are describing. I need to investigate further.

**Comment by pykong** [comment]
Done. Dunno whar prevented the resolution of the conflicts in the first place.

As a side-note:
Maybe you want to wait with the release for 2 weeks this would allow me to address #55 and ä57.

**Comment by pykong** [comment]
@jpscaletti Can we move this PR forward? Please have a look!

File: copier/config/factory.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+from typing import Tuple
+
+from ..types import AnyByStrDict, OptAnyByStrDict, OptBool, OptStrSeq
+from .objects import DEFAULT_DATA, ConfigData, EnvOps, Flags
+from .user_data import load_config_data, query_user_data
+
+__all__ = "make_config"
```
Comment (jpscaletti) [inline copier/config/factory.py]: This is invalid. `__all__` must be a list (any iterable actually), so
```python
__all__ = "make_config"
```
is like wrtiting 
```python
__all__ = ["m", "a", "k", "e", "_", "c", "", "o", "n", "f", "i", "g"]
```
This should be `__all__ = ["make_config"]`
See: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

**Comment by pykong** [comment]
Great! Thanks for merging!

File: copier/config/factory.py (hunk 1 above)
Comment (pykong) [reply copier/config/factory.py]: Thanks for your compliments. Your code was a joy to work with, especially due to the extensive test coverage provided.

Could `black` be the culprit here?
I do not remember having made this change manually.

Comment (pykong) [reply copier/config/factory.py]: @jpscaletti
Could you maybe add some regression tests to ensure the integrity of the `__all__` dunders?

**Pull request merged**